### PR TITLE
Code4rena style and gas optimization

### PR DESCRIPTION
**Pull request by grandizzy**
Style:
- [N‑11] | Lines are too long
- [N‑12] | Typos
- [N‑13] | NatSpec @param is missing 
- [N‑14] | NatSpec @return argument is missing

Gas Optimizations:
- [G‑02] Using storage instead of memory for structs/arrays saves gas
- [G‑03] State variables should be cached in stack variables rather than re-reading them from storage
- [G‑04] Multiple accesses of a mapping/array should use a local variable cache
- [G‑07] <array>.length should not be looked up in every loop of a for-loop

* Gas pre
```
| Function Name                               | min             | avg    | median | max    | # calls |
| claimDelegateReward                         | 1407            | 55631  | 58106  | 64906  | 90      |
| executeExtraordinary                        | 7378            | 44134  | 13238  | 95408  | 5       |
| executeStandard                             | 9009            | 23123  | 9606   | 47334  | 5       |
| findMechanismOfProposal                     | 613             | 2043   | 799    | 4719   | 3       |
| fundTreasury                                | 7882            | 41558  | 44522  | 66992  | 32      |
| fundingVote                                 | 1584            | 106694 | 106065 | 409114 | 100     |
| getDelegateReward                           | 2979            | 3469   | 2979   | 5021   | 5       |
| getDistributionId                           | 374             | 465    | 374    | 2374   | 305     |
| getDistributionPeriodInfo                   | 1046            | 1582   | 1046   | 5046   | 82      |
| getExtraordinaryProposalInfo                | 980             | 980    | 980    | 980    | 3       |
| getExtraordinaryProposalSucceeded           | 2400            | 2821   | 2897   | 3167   | 3       |
| getFundedProposalSlate                      | 1133            | 1309   | 1368   | 1368   | 4       |
| getFundingVotesCast                         | 1577            | 2317   | 2317   | 3057   | 2       |
| getMinimumThresholdPercentage               | 471             | 723    | 471    | 2471   | 9       |
| getProposalInfo                             | 980             | 980    | 980    | 980    | 107     |
| getSlateHash                                | 872             | 880    | 884    | 884    | 3       |
| getSliceOfNonTreasury                       | 1738            | 5988   | 5988   | 10238  | 2       |
| getSliceOfTreasury                          | 783             | 1783   | 1783   | 2783   | 2       |
| getTopTenProposals                          | 1658            | 2743   | 2363   | 3304   | 13      |
| getVoterInfo                                | 1002            | 1002   | 1002   | 1002   | 5       |
| getVotesExtraordinary                       | 764             | 7469   | 7782   | 8204   | 117     |
| getVotesFunding                             | 2389            | 6432   | 2797   | 11619  | 15      |
| getVotesScreening                           | 5032            | 5064   | 5032   | 7032   | 192     |
| hasVotedExtraordinary                       | 683             | 1683   | 1683   | 2683   | 2       |
| hashProposal                                | 3843            | 3843   | 3843   | 3843   | 79      |
| proposeExtraordinary                        | 4731            | 55478  | 88452  | 88622  | 12      |
| proposeStandard                             | 4709            | 78784  | 82665  | 85004  | 81      |
| screeningVote                               | 1847            | 51270  | 47547  | 395996 | 119     |
| screeningVotesCast                          | 663             | 1329   | 663    | 2663   | 3       |
| startNewDistributionPeriod                  | 617             | 49323  | 53165  | 53165  | 16      |
| state                                       | 1569            | 4615   | 4008   | 7495   | 12      |
| treasury                                    | 396             | 396    | 396    | 396    | 6       |
| updateSlate                                 | 971             | 56228  | 31579  | 318349 | 14      |
| voteExtraordinary                           | 565             | 27563  | 27559  | 31559  | 114     |
```

* Gas post
```
| Function Name                               | min             | avg    | median | max    | # calls |
| claimDelegateReward                         | 1407            | 56803  | 58106  | 64906  | 171     |
| executeExtraordinary                        | 7378            | 43278  | 26160  | 95362  | 6       |
| executeStandard                             | 9009            | 23102  | 9606   | 47281  | 5       |
| findMechanismOfProposal                     | 613             | 2043   | 799    | 4719   | 3       |
| fundTreasury                                | 7882            | 41558  | 44522  | 66992  | 32      |
| fundingVote                                 | 1584            | 106384 | 106065 | 409114 | 181     |
| getDelegateReward                           | 2979            | 3469   | 2979   | 5021   | 5       |
| getDistributionId                           | 374             | 436    | 374    | 2374   | 450     |
| getDistributionPeriodInfo                   | 1046            | 1722   | 1046   | 5046   | 65      |
| getExtraordinaryProposalInfo                | 980             | 980    | 980    | 980    | 3       |
| getExtraordinaryProposalSucceeded           | 2407            | 2796   | 2904   | 3077   | 3       |
| getFundedProposalSlate                      | 1133            | 1309   | 1368   | 1368   | 4       |
| getFundingVotesCast                         | 1577            | 2317   | 2317   | 3057   | 2       |
| getMinimumThresholdPercentage               | 478             | 695    | 478    | 2478   | 10      |
| getProposalInfo                             | 980             | 980    | 980    | 980    | 107     |
| getSlateHash                                | 872             | 880    | 884    | 884    | 3       |
| getSliceOfNonTreasury                       | 1738            | 5988   | 5988   | 10238  | 2       |
| getSliceOfTreasury                          | 783             | 1783   | 1783   | 2783   | 2       |
| getTopTenProposals                          | 1658            | 2743   | 2363   | 3304   | 13      |
| getVoterInfo                                | 1002            | 1002   | 1002   | 1002   | 5       |
| getVotesExtraordinary                       | 764             | 6904   | 7366   | 8204   | 281     |
| getVotesFunding                             | 2389            | 6432   | 2797   | 11619  | 15      |
| getVotesScreening                           | 5032            | 5049   | 5032   | 7032   | 354     |
| hasVotedExtraordinary                       | 683             | 1683   | 1683   | 2683   | 2       |
| hashProposal                                | 3843            | 3843   | 3843   | 3843   | 63      |
| proposeExtraordinary                        | 4731            | 57540  | 88283  | 88622  | 13      |
| proposeStandard                             | 4709            | 77261  | 82879  | 84879  | 64      |
| screeningVote                               | 1538            | 44734  | 39393  | 396654 | 200     |
| screeningVotesCast                          | 663             | 1329   | 663    | 2663   | 3       |
| startNewDistributionPeriod                  | 617             | 49316  | 53165  | 53165  | 16      |
| state                                       | 1569            | 4719   | 4019   | 7502   | 13      |
| treasury                                    | 396             | 396    | 396    | 396    | 7       |
| updateSlate                                 | 971             | 56230  | 31581  | 318349 | 14      |
| voteExtraordinary                           | 565             | 27697  | 27559  | 31559  | 278     |
```

### CODE REVIEW

**Review by MikeHathaway — COMMENTED** [review]
Interesting that the gas cost decreases with storage. Can you add a gas snapshot? Generally LGTM

File: src/grants/base/ExtraordinaryFunding.sol [hunk 1]
```diff
@@ -204,14 +221,10 @@ abstract contract ExtraordinaryFunding is Funding, IExtraordinaryFunding {
...
-            return 0.5 * 1e18;
-        }
-        // minimum threshold increases according to the number of funded EFM proposals
-        else {
-            return 0.5 * 1e18 + (_fundedExtraordinaryProposals.length * (0.05 * 1e18));
-        }
+        uint256 noOfProposals = _fundedExtraordinaryProposals.length;
+        // default minimum threshold (when no proposal) is 50%
+        // if there are proposals then minimum threshold increases according to the number of funded EFM proposals
+        return (noOfProposals == 0) ? 0.5 * 1e18 : 0.5 * 1e18 + (noOfProposals * (0.05 * 1e18));
```
Comment (MikeHathaway) [inline src/grants/base/ExtraordinaryFunding.sol]: Does the ternary save on gas? Think its otherwise much harder to read as a ternary

**Comment by grandizzy** [comment]
> Interesting that the gas cost decreases with storage. Can you add a gas snapshot? Generally LGTM

yeah, will add gas snapshots. It actually make sense as those places were loading entire struct members with `memory` as with `storage` we're only loading what we need from struct

File: src/grants/base/ExtraordinaryFunding.sol (hunk 1 above)
Comment (grandizzy) [reply src/grants/base/ExtraordinaryFunding.sol]: nope, it doesn't, will put it back as non ternary and keep the `noOfProposals` load once

Comment (grandizzy) [reply src/grants/base/ExtraordinaryFunding.sol]: reverted with https://github.com/ajna-finance/ecosystem-coordination/pull/79/commits/d1f27f11d9f240f3ed635c64aa4b79acbf142ffb

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Review by prat-gpt — APPROVED** [review]
LGTM!

**Pull request merged**